Make it work, add ability to mark current state

haffla opened the pull request:
Hello, I believe this actually didn't work anymore. I kept getting `ArgumentError` for `Stateoscope::Adapter::Base#initialize`.  Is it still maintained? 

I made it work again and added some sugar. You can now configure the output format via `options` and also pass a `current_state` which node will be coloured green. I believe this can be very helpful.

padde commented:
Hey @haffla thanks for your contribution! I must have overlooked your PR, sorry it took so long. I will take a look soon and release a new version. First I want to fix the CI build and update some of the development dependencies though.

haffla commented:
cool I'll update the code to tame the houndci-bot

padde commented:
@haffla build is working again, please merge/rebase current `master`.

haffla commented:
I've fixed the failing tests, not sure why I had overlooked the test suite.

They mostly failed because now that `Base` is not a struct anymore, but a class, it's constructor parameters are actually required. The overall api of the gem did not change.

Also would it make sense to add Rubocop and enforce a style + add check to Travis?

I will update the Readme, too.

padde commented on spec/lib/stateoscope/tasks/stateoscope_rake_spec.rb:
```diff
@@ -41,5 +41,14 @@
+        rake['stateoscope:visualize'].invoke('DummyClass', 'foo', 'bar')
+        expect(stateoscope).to have_received(:visualize).with(DummyClass, state_machine_name: 'foo', current_state: 'bar')
+      end
+    end
```
It would be helpful if you could add this variant to the README for documentation purposes.

padde commented on lib/tasks/stateoscope.rake:
```diff
@@ -1,8 +1,10 @@
-    Stateoscope.visualize(klass, state_machine_name: args[:state_machine_name])
+    # class not needed in args
+    args = args.reject { |k, _| k == :class }.to_h
+    Stateoscope.visualize(klass, args)
```
I believe this can be improved by using `Hash#delete`. It is more concise and we don't need to repeat the`:class` key:

    klass = args.delete(:class).classify.constantize
    Stateoscope.visualize(klass, args)

haffla replied:
yea I tried `Hash#delete` but didn't work. I actually just found out that it's not a `Hash` but a `Rake::TaskArguments`. I think it would actually be best to have `Stateoscope.visualize` just take one argument (a `Hash`), but I'm not gonna change that now.

padde commented on lib/tasks/stateoscope.rake:
```diff
@@ -1,8 +1,10 @@
-  task :visualize, [:class, :state_machine_name] => :environment do |t, args|
-    fail ArgumentError, 'missing required argument <class>' unless args[:class].present?
+  task :visualize, %i[class state_machine_name current_state] => :environment do |_, args|
+    raise ArgumentError, 'missing required argument <class>' if args[:class].nil?
```
`if x.nil?` is not exactly the same as `unless x.present?` - it now allows blank strings. I would prefer that a blank string still trigger the exception here. I think `if x.blank?` will do the trick.

haffla replied:
Yes sorry not sure why I changed that. I'm working with vscode and a rubocop extension and probably it complained about that part because `unless something` can always be rewritten with `if the opposite`. Sure `nil?` is not the opposite of `present?` I know. 

By the way `blank?`, `present?`, `classify`, `constantize` are all Rails. Have you tried to use the gem outside of Rails?

padde replied:
Yes I thought about that as well while writing this comment but at the moment there is a dependency on activesupport in the gemfile anyway. We might want to get rid of that in the future, see #7

padde commented on lib/stateoscope/visualizer.rb:
```diff
@@ -10,8 +10,11 @@ def parse_graph
+      options = {}
+      options[format] = "#{filename}.#{format}"
+      @viz.output(options)
+      "#{filename}.#{format}"
```
Why is this line necessary?

haffla replied:
I'm using Stateoscope not through rake but dynamically in a Rails project. In one of my controllers I do

```
filepath = Stateoscope.visualize(
        BaustarterState,
        current_state: @user.baustarter_state.state,
        dir: Rails.root.join("public"),
        format: "png"
      )

# return filepath to client
```

Anyway since the filename is created from a timestamp in `Stateoscope.filename_for` it would be nice to return the final filename to the caller, otherwise they have no idea where the just created file is located.

haffla replied:
ok yea I removed it from the visualizer and just return the filename from `visualize` method. Better.

padde commented on lib/stateoscope/visualizer.rb:
```diff
@@ -10,8 +10,11 @@ def parse_graph
+    def output(filename, format)
+      options = {}
+      options[format] = "#{filename}.#{format}"
+      @viz.output(options)
```
This can be boiled down to

```
@viz.output(format => "#{filename}.#{format}")
```

However, I don't like how the responsibility of generating the filename is now split across two classes. I'd rather add a `format` argument to `Stateoscope.filename_for` so we can simply use the full filename here. What do you think?

BTW I realize it was that way before, my bad 😉

haffla replied:
you are right. gonna change it

haffla commented:
@padde thanks for the feedback. Right, there were some things that could be improved. I refactored the code a bit, added more documentation and specs.

haffla commented:
Great! Will follow the progress! :)

padde commented:
@haffla I've *finally* released v0.1.2 😄
